File: ai_service/app/models/ai_input_video.py
"""
SQLAlchemy model for AI Input Video indexing.
Tracks user-uploaded videos that are being processed (indexed) for metadata
extraction — transcript, visual data, speaker foreground, etc.
"""
from __future__ import annotations

from datetime import datetime
from typing import Dict, Any
from uuid import uuid4

from sqlalchemy import Column, String, Text, Float, Integer, DateTime, Index, text
from sqlalchemy.dialects.postgresql import UUID, JSONB

from .ai_gen_video import Base  # reuse the same declarative Base


class AiInputVideo(Base):
    """
    Tracks video input indexing lifecycle.

    Status flow: PENDING → QUEUED → PROCESSING → COMPLETED | FAILED
    """
    __tablename__ = "ai_input_videos"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4,
                server_default=text("gen_random_uuid()"))
    institute_id = Column(Text, nullable=False, index=True)
    name = Column(String(255), nullable=False)
    mode = Column(String(20), nullable=False)  # 'podcast' or 'demo'
    status = Column(String(50), nullable=False, default="PENDING", index=True)
    source_url = Column(Text, nullable=False)

    duration_seconds = Column(Float, nullable=True)
    resolution = Column(Text, nullable=True)

    # Indexing output URLs (filled on completion)
    context_json_url = Column(Text, nullable=True)
    spatial_db_url = Column(Text, nullable=True)
    assets_urls = Column(JSONB, nullable=False, default=dict,
                         server_default=text("'{}'::jsonb"))

    # Render worker job tracking
    render_job_id = Column(String(255), nullable=True)
    progress = Column(Integer, nullable=False, default=0, server_default=text("0"))
    error_message = Column(Text, nullable=True)

    # Flexible metadata
    extra_metadata = Column('metadata', JSONB, nullable=False, default=dict,
                            server_default=text("'{}'::jsonb"))

    created_by_user_id = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow,
                        onupdate=datetime.utcnow)

    def __repr__(self) -> str:
        return f"<AiInputVideo(id={self.id}, name={self.name}, status={self.status})>"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "institute_id": self.institute_id,
            "name": self.name,
            "mode": self.mode,
            "status": self.status,
            "source_url": self.source_url,
            "duration_seconds": self.duration_seconds,
            "resolution": self.resolution,
            "context_json_url": self.context_json_url,
            "spatial_db_url": self.spatial_db_url,
            "assets_urls": self.assets_urls or {},
            "render_job_id": self.render_job_id,
            "progress": self.progress or 0,
            "error_message": self.error_message,
            "metadata": self.extra_metadata or {},
            "created_by_user_id": self.created_by_user_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }


Index("idx_aiv_institute", AiInputVideo.institute_id)
Index("idx_aiv_status", AiInputVideo.status)

File: admin_core_service/src/main/java/vacademy/io/admin_core_service/features/ai_input_video/entity/AiInputVideo.java
package vacademy.io.admin_core_service.features.ai_input_video.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.UuidGenerator;
import org.hibernate.type.SqlTypes;

import java.sql.Timestamp;

@Entity
@Getter
@Setter
@NoArgsConstructor
@Table(name = "ai_input_videos")
public class AiInputVideo {

    @Id
    @UuidGenerator
    private String id;

    @Column(name = "institute_id", nullable = false)
    private String instituteId;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "mode", nullable = false, length = 20)
    private String mode;

    @Column(name = "status", nullable = false, length = 50)
    private String status = "PENDING";

    @Column(name = "source_url", nullable = false, columnDefinition = "TEXT")
    private String sourceUrl;

    @Column(name = "duration_seconds")
    private Float durationSeconds;

    @Column(name = "resolution")
    private String resolution;

    @Column(name = "context_json_url", columnDefinition = "TEXT")
    private String contextJsonUrl;

    @Column(name = "spatial_db_url", columnDefinition = "TEXT")
    private String spatialDbUrl;

    @Column(name = "assets_urls", columnDefinition = "jsonb")
    @JdbcTypeCode(SqlTypes.JSON)
    private String assetsUrls;

    @Column(name = "render_job_id")
    private String renderJobId;

    @Column(name = "progress")
    private Integer progress = 0;

    @Column(name = "error_message", columnDefinition = "TEXT")
    private String errorMessage;

    @Column(name = "metadata", columnDefinition = "jsonb")
    @JdbcTypeCode(SqlTypes.JSON)
    private String metadata;

    @Column(name = "created_by_user_id")
    private String createdByUserId;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Timestamp createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Timestamp updatedAt;
}
